<center>
    <img src="https://806230.smushcdn.com/1739487/wp-content/uploads/2021/07/Linear-vs-Multiple-Regression-02-min-768x402.png?lossy=0&strip=1&webp=1" width="400" alt="Regression"  />
</center>

# Multiple Linear Regression

a statistical technique that uses several independent variables to predict the outcome of a dependent variable

## Objectives

*   Use scikit-learn to implement Multiple Linear Regression
*   Create a model, train it, test it and use the model

In [2]:
import matplotlib.pyplot as plt
import pandas as pd
import pylab as pl
import numpy as np
%matplotlib inline

In [3]:
!wget -O FuelConsumption.csv https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv

--2022-01-02 18:38:23--  https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv
Resolving cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)... 198.23.119.245
Connecting to cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)|198.23.119.245|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 72629 (71K) [text/csv]
Saving to: ‘FuelConsumption.csv’

FuelConsumption.csv 100%[===================>]  70.93K  --.-KB/s    in 0.005s  

2022-01-02 18:38:24 (15.3 MB/s) - ‘FuelConsumption.csv’ saved [72629/72629]



In [4]:
df = pd.read_csv("FuelConsumption.csv")
# take a look at the dataset
df.head()

MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   
4       2014  ACURA     RDX AWD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0          AS5        Z                   9.9                  6.7   
1           M6        Z                  11.2                  7.7   
2          AV7        Z                   6.0                  5.8   
3          AS6        Z                  12.7                  9.1   
4          AS6        Z                  12.1                  8.7   

   FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                   8.5                        33           196  
1                   9.6                        29           221  
2                   5.9                        48           136  
3                  11.1                        25           255  
4                  10.6                        27           244

In [7]:
msk = np.random.rand(len(df)) < 0.8
train = df[msk]
test = df[~msk]

In [8]:
from sklearn import linear_model
regr = linear_model.LinearRegression()
x = np.asanyarray(train[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB']])
y = np.asanyarray(train[['CO2EMISSIONS']])
regr.fit (x, y)
# The coefficients
print ('Coefficients: ', regr.coef_)

Coefficients:  [[10.54026675  8.1604935   9.52664464]]


In [9]:
y_hat= regr.predict(test[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB']])
x = np.asanyarray(test[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB']])
y = np.asanyarray(test[['CO2EMISSIONS']])
print("Residual sum of squares: %.2f"
      % np.mean((y_hat - y) ** 2))

# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % regr.score(x, y))

Residual sum of squares: 569.23
Variance score: 0.84


### What is the better group of independent variables

In [10]:
a = np.asanyarray(train[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB_MPG']])
b = np.asanyarray(train[['CO2EMISSIONS']])
regr.fit (a, b)
# The coefficients
print ('Coefficients: ', regr.coef_)

Coefficients:  [[ 9.89270488  7.42196009 -4.85585857]]


In [11]:
y_hats= regr.predict(test[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB_MPG']])
c = np.asanyarray(test[['ENGINESIZE','CYLINDERS','FUELCONSUMPTION_COMB_MPG']])
d = np.asanyarray(test[['CO2EMISSIONS']])
print("Residual sum of squares: %.2f"
      % np.mean((y_hats - d) ** 2))

# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % regr.score(c, d))

Residual sum of squares: 457.74
Variance score: 0.87


### Conclusion 

ENGINESIZE,CYLINDERS,FUELCONSUMPTION_COMB_MPG - model has a higher variance of 0.87 and closer to perfect score of 1